### ISSUE TITLE
Diff pathological case

### ISSUE THREAD

**Issue by fasterthanlime** [itchio/butler#138]
**TL:DR: In some rare cases, diffing becomes extremely slow because it's computing too many MD5 hashes**

Rsync diffing works like that: it scans the new data using a 64KB window. In each step it:
  * Computes a weak hash (really fast, especially if we advanced 1 byte only)
  * If this weak hash is in the hash library (signature of the previous build)
    * It computes a "strong" hash (md5)
    * Compares it to the entry stored in the library
    * If it's a match, then we can re-use that block (and we advanced by 64KB)

I've encountered a case where:

  * the weak hash remains `0` for a long while,
  * there's entries in the hash library for `0`, so it computes MD5 hashes every time
  * it only advances by 1 byte each time

So it computes the strong hash for:

```
[ this part ]xxxxxxxx
x[ this part ]xxxxxxx
xx[ this part ]xxxxxx
xxx[ this part ]xxxxx
```

...and so on. Eventually, it doesn't find much to re-use (the strong hash is very rarely a match).

When pushing such a build, it appears that the process is stuck (but it's not, it's using up 100% CPU to compute tons and tons of MD5 hashes).

---

I've designed a workaround that does not attempt to compute a strong hash if:
  * We just computed a rolling weak hash
  * It has the exact same value as the previous weak hash

This means we'll fail to identify some blocks to be re-used, on occasion (so we'll make slightly worse "simple" patches - rediff patches are unaffected by this change).

But look at the performance difference: top is with workaround, bottom is without:

![pathological-case](https://user-images.githubusercontent.com/7998310/35860103-0057f32c-0b43-11e8-9bf2-b56dd15d8d7e.png)

---

I'm going to ship a new stable version of butler today with this patch, I just wanted to have this issue here for the record :shipit:
